### PR TITLE
Clean up docs and add tests

### PR DESCRIPTION
**Pull request by forest**
- [x] add dependabot
- [x] refactor CI
- [x] fix dialyzer errors
- [x] update docs
- [x] general code cleanup

### CODE REVIEW

File: lib/snowpack/query.ex [hunk 1]
```diff
@@ -35,9 +35,7 @@ defmodule Snowpack.Query do
             opts :: Keyword.t()
           ) :: [Type.param()]
     def encode(_query, params, opts) do
-      params
-      |> List.flatten()
```
Comment (forest) [inline lib/snowpack/query.ex:39]: Don't want to flatten a param of type List. Not that List params are supported yet, but this just makes no sense.

`[1, 2, ["a", "b"], 5]` --> `[1, 2, "a", "b", 5]` is not what we want.

File: lib/snowpack/type.ex [hunk 2]
```diff
@@ -137,12 +138,33 @@ defmodule Snowpack.Type do
...
+    Enum.find_value(parse_funcs, value, fn func ->
+      case Kernel.apply(func, [value]) do
+        {integer, ""} -> integer
+        {:ok, result} -> result
+        _ -> false
+      end
+    end)
+  end
+
+  def decode(value, _opts), do: value
```
Comment (forest) [inline lib/snowpack/type.ex:167]: If the type parsing fails for some reason these `decode/2` are called by `DBConnection` and can make a best attempt for types coming from Erlang `:odbc`.

File: lib/snowpack/type_parser.ex [hunk 3]
```diff
@@ -40,11 +42,16 @@ defmodule Snowpack.TypeParser do
...
 
+  defp parse({:float, data}) when is_float(data), do: data
+
   defp parse({:float, data}) when is_binary(data) do
-    {float, ""} = Float.parse(data)
-    Decimal.from_float(float)
+    case Float.parse(data) do
+      {float, _rest} -> Decimal.from_float(float)
+      :error -> return_raw(:float, data, :float_parse_error)
+    end
```
Comment (forest) [inline lib/snowpack/type_parser.ex:51]: Handle errors in parsing. Log and return the original value. Probably, better than blowing up.

File: test/query_test.exs [hunk 4]
```diff
@@ -29,16 +29,10 @@ defmodule QueryTest do
...
-      #    credo:disable-for-next-line Credo.Check.Design.TagTODO
-      #    TODO: Get working test for OBJECT type
-      #    object = %{a: 1, b: 'BBBB'}
-      #    assert [[^object]] =
-      #             query("SELECT object_construct('a',1,'b','BBBB') as obj", [])
     end
 
     test "long number param", context do
-      assert [["123456789012345678901"]] = query("SELECT ?", [123_456_789_012_345_678_901])
+      assert [[123_456_789_012_345_678_901]] = query("SELECT ?", [123_456_789_012_345_678_901])
```
Comment (forest) [inline test/query_test.exs:35]: Fixed large number parsing to return an Integer instead of String.

Comment (forest) [reply test/query_test.exs:35]: @sarahdasko this is probably a breaking change for us.

**Comment by raylas** [comment]
CI stuff looks good.

If we like this pattern then let's  make sure to back-fill [phoenix-liveview-template](https://github.com/HGData/phoenix-liveview-template).

Also, the first time I've seen this:
Screen Shot 2022-05-03 at 8 43 57 AM

Something we maybe want to consider in the future, HgCd or not.

https://docs.github.com/en/rest/deployments/deployments#create-a-deployment

**Comment by forest** [comment]
@raylas I'm not sure I like the multiple workflows pattern. In Actions tab you have 3 workflows running which is fine but it doesn't show the nice graph of dependent jobs. In this case, they are not dependent and run in parallel, but you also can't make them dependent this way.

Thoughts?

File: test/query_test.exs (hunk 4 above)
Comment (sarahdasko) [reply test/query_test.exs:35]: possibly, we'll have to run regression tests on all of our apps to ensure either way.

**Review by sarahdasko — APPROVED** [review]
looks great

File: lib/snowpack/protocol.ex [hunk 5]
```diff
@@ -32,7 +32,8 @@ defmodule Snowpack.Protocol do
 
   @type opts :: Keyword.t()
   @type query :: Snowpack.Query.t()
-  @type params :: [{:odbc.odbc_data_type(), :odbc.value()}]
+  @type value :: :null | term()
+  @type params :: [{atom(), [value()]}] | [{atom(), atom(), [value()]}]
```
Comment (aschreifels) [inline lib/snowpack/protocol.ex:36]: Would it be worthwhile to create 2 custom types for these two different versions of params?

**Review by aschreifels — APPROVED** [review]

File: lib/snowpack/protocol.ex (hunk 5 above)
Comment (forest) [reply lib/snowpack/protocol.ex:36]: I wish dialyzer would be able to use the ones from `:odbc`. I'll update to be clear where these are coming from.

**Review by aschreifels — APPROVED** [review]

**Pull request merged**